src/parser/input_utils.rs

```diff
@@ -0,0 +1,166 @@
+use crate::parser::utils::{string_to_node, NixNode};
+use crate::SmlStr;
+use rnix::{NixLanguage, SyntaxKind::*};
+use rowan::{api::SyntaxNode, GreenNode, GreenNodeBuilder, GreenToken, Language, NodeOrToken};
+use std::string::ToString;
+
+#[derive(Debug, Clone, Eq, PartialEq, Default)]
+pub struct Input {
+    pub(crate) name: Option<SyntaxStructure>,
+    pub(crate) url: Option<SyntaxStructure>,
+    pub(crate) is_flake: Option<SyntaxStructure>,
+}
+
+pub fn wrap_root(node: GreenNode) -> NixNode {
+    let kind: rowan::SyntaxKind = NixLanguage::kind_to_raw(NODE_ROOT);
+    let root = GreenNode::new(kind, vec![NodeOrToken::Node(node)]);
+    SyntaxNode::new_root(root)
+}
+
+pub fn new_string(s: String) -> GreenNode {
+    let mut node = GreenNodeBuilder::new();
+    let kind: rowan::SyntaxKind = NixLanguage::kind_to_raw(NODE_STRING);
+    node.start_node(kind);
+    let start_string_kind: rowan::SyntaxKind = NixLanguage::kind_to_raw(TOKEN_STRING_START);
+    node.token(start_string_kind, "\"");
+    let string_content: rowan::SyntaxKind = NixLanguage::kind_to_raw(TOKEN_STRING_CONTENT);
+    node.token(string_content, &s);
+    let end_string_kind: rowan::SyntaxKind = NixLanguage::kind_to_raw(TOKEN_STRING_END);
+    node.token(end_string_kind, "\"");
+    node.finish_node();
+    node.finish()
+}
+
+/// creates a bool literal. Example of what
+/// this should look like structurally:
+///
+/// NODE_IDENT {
+///   TOKEN_IDENT("true")
+/// }
+///TOKEN_SEMICOLON(";") 23..
+pub fn new_bool_literal(b: bool) -> GreenNode {
+    let mut node = GreenNodeBuilder::new();
+    let token_ident_kind: rowan::SyntaxKind = NixLanguage::kind_to_raw(TOKEN_IDENT);
+    let node_ident_kind: rowan::SyntaxKind = NixLanguage::kind_to_raw(NODE_IDENT);
+    node.start_node(node_ident_kind);
+    node.token(token_ident_kind, b.to_string().as_str());
+    node.finish_node();
+    node.finish()
+}
+
+pub fn new_key(s: String) -> GreenNode {
+    let kind: rowan::SyntaxKind = NixLanguage::kind_to_raw(NODE_KEY);
+    let children = vec![NodeOrToken::Node(new_string(s))];
+    GreenNode::new(kind, children)
+}
+
+pub fn new_key_value(key: GreenNode, value: GreenNode) -> GreenNode {
+    let kind = NixLanguage::kind_to_raw(NODE_KEY_VALUE);
+    let assign_kind = NixLanguage::kind_to_raw(TOKEN_ASSIGN);
+    let whitespace_kind = NixLanguage::kind_to_raw(TOKEN_WHITESPACE);
+    let semicolon_kind = NixLanguage::kind_to_raw(TOKEN_SEMICOLON);
+    let children = vec![
+        NodeOrToken::Node(key),
+        NodeOrToken::Token(GreenToken::new(whitespace_kind, " ")),
+        NodeOrToken::Token(GreenToken::new(assign_kind, "=")),
+        NodeOrToken::Token(GreenToken::new(whitespace_kind, " ")),
+        NodeOrToken::Node(value),
+        NodeOrToken::Token(GreenToken::new(semicolon_kind, ";")),
+        NodeOrToken::Token(GreenToken::new(whitespace_kind, "\n")),
+    ];
```

Comment:
```suggestion
    use NixLanguage::kind_to_raw;
    use NodeOrToken::{Node, Token};
    let kind = kind_to_raw(NODE_KEY_VALUE);
    let assign_kind = kind_to_raw(TOKEN_ASSIGN);
    let whitespace_kind = kind_to_raw(TOKEN_WHITESPACE);
    let semicolon_kind = kind_to_raw(TOKEN_SEMICOLON);
    let children = vec![
        Node(key),
        Token(GreenToken::new(whitespace_kind, " ")),
        Token(GreenToken::new(assign_kind, "=")),
        Token(GreenToken::new(whitespace_kind, " ")),
        Node(value),
        Token(GreenToken::new(semicolon_kind, ";")),
        Token(GreenToken::new(whitespace_kind, "\n")),
    ];
```

Reply:
Can't import NixLanguage::kind_to_raw. Not sure why....

Reply:
Oh... (that might be the case because the method is defined in a trait impl)
Maybe create a small wrapper for it then:
```rust
#[inline(always)]
fn kind_to_raw(x: rnix::SyntaxKind) -> rowan::SyntaxKind {
    NixLanguage::kind_to_raw(x)
}
```